drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,612 @@
+
+
+static const struct uart_driver_api uart_nrfx_uarte_driver_api = {
+	.poll_in		= uarte_nrfx_poll_in,
+	.poll_out		= uarte_nrfx_poll_out,
+	.err_check		= uarte_nrfx_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill		= uarte_nrfx_fifo_fill,
+	.fifo_read		= uarte_nrfx_fifo_read,
+	.irq_tx_enable		= uarte_nrfx_irq_tx_enable,
+	.irq_tx_disable		= uarte_nrfx_irq_tx_disable,
+	.irq_tx_ready		= uarte_nrfx_irq_tx_ready_complete,
+	.irq_rx_enable		= uarte_nrfx_irq_rx_enable,
+	.irq_rx_disable		= uarte_nrfx_irq_rx_disable,
+	.irq_tx_complete	= uarte_nrfx_irq_tx_ready_complete,
+	.irq_rx_ready		= uarte_nrfx_irq_rx_ready,
+	.irq_err_enable		= uarte_nrfx_irq_err_enable,
+	.irq_err_disable	= uarte_nrfx_irq_err_disable,
+	.irq_is_pending		= uarte_nrfx_irq_is_pending,
+	.irq_update		= uarte_nrx_irq_update,
+	.irq_callback_set	= uarte_nrfx_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static int uarte_instance_init(struct device *dev,
+			       const struct uarte_init_config *config,
+			       u8_t interrupts_active)
+{
+	int err;
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	nrf_gpio_pin_write(config->pseltxd, 1);
+	nrf_gpio_cfg_output(config->pseltxd);
+
+	nrf_gpio_cfg_input(config->pselrxd, NRF_GPIO_PIN_NOPULL);
+
+	nrf_uarte_txrx_pins_set(uarte,
+				config->pseltxd,
+				config->pselrxd);
+
+	if (config->hwfc == NRF_UARTE_HWFC_ENABLED) {
+		nrf_gpio_pin_write(config->pselrts, 1);
+		nrf_gpio_cfg_output(config->pselrts);
+
+		nrf_gpio_cfg_input(config->pselcts, NRF_GPIO_PIN_NOPULL);
+
+		nrf_uarte_hwfc_pins_set(uarte,
+					config->pselrts,
+					config->pselcts);
+	}
+
+	/* Configure flow control and parity checking */
+	nrf_uarte_configure(uarte,
+			    config->parity,
+			    config->hwfc);
+
+	err = baudrate_set(dev, config->baudrate);
+	if (err) {
+		return err;
+	}
+
+	/* Enable receiver and transmitter */
+	nrf_uarte_enable(uarte);
+
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
+
+	nrf_uarte_rx_buffer_set(uarte, &data->rx_data, 1);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+
+#if CONFIG_UART_INTERRUPT_DRIVEN
+	if (interrupts_active) {
+
+		irq_enable(NRFX_IRQ_NUMBER_GET(uarte));
+
+		/* Set ENDTX event by requesting fake (zero-length) transfer.
+		 * Pointer to RAM variable (data->tx_buffer) is set because
+		 * otherwise such operation may result in HardFault or RAM
+		 * corruption.
+		 */
+		nrf_uarte_tx_buffer_set(uarte, data->tx_buffer, 0);
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
+
+		/* switch off transmitter to save an energy */
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+	}
+#endif
+
+	return 0;
+}
+
+#define UART_NRF_UARTE_DEVICE(idx)					      \
+	DEVICE_DECLARE(uart_nrfx_uarte##idx);				      \
+	static u8_t uarte##idx##_tx_buffer[UART##idx##_BUFF_SIZE];	      \
+	static struct uarte_nrfx_data uarte_##idx##_data = {		      \
+		.tx_buffer = uarte##idx##_tx_buffer			      \
```

Comment:
This won't compile when `CONFIG_UART_INTERRUPT_DRIVEN` is not defined.
You'll need to use more macros here.